src/Service/Indexer/InternalResourceIndexer.php

```diff
@@ -0,0 +1,366 @@
+        string $processId,
+        IndexerParameter $parameter,
+        string $index,
+        array $pathList
+    ): void {
+        $offset = 0;
+        $successCount = 0;
+
+        while (true) {
+            $indexedCount = $this->indexChunks(
+                $processId,
+                $index,
+                $pathList,
+                $offset,
+                $parameter->chunkSize
+            );
+            if ($indexedCount === false) {
+                break;
+            }
+            $successCount += $indexedCount;
+            $offset += $parameter->chunkSize;
+            gc_collect_cycles();
+        }
+
+        if (
+            $parameter->cleanupThreshold > 0 &&
+            $successCount >= $parameter->cleanupThreshold
+        ) {
+            $this->indexService->deleteExcludingProcessId(
+                $index,
+                $this->source,
+                $processId
+            );
+        }
+        $this->indexService->commit($index);
+    }
+
+    /**
+     * For performance reasons, not every resource is indexed individually,
+     * but the index documents are first generated from several resources.
+     * These are then passed to Solr for indexing via a request. These
+     * methods accept a chunk with all paths that are to be indexed via a
+     * request.
+     *
+     * @param string[] $pathList
+     */
+    private function indexChunks(
+        string $processId,
+        string $solrCore,
+        array $pathList,
+        int $offset,
+        int $length
+    ): int|false {
+        $resourceList = $this->loadResources(
+            $pathList,
+            $offset,
+            $length
+        );
+        if ($resourceList === false) {
+            return false;
+        }
+        if ($this->aborter->shouldAborted($solrCore)) {
+            $this->aborter->aborted($solrCore);
+            $this->indexerProgressHandler->abort();
+            return false;
+        }
+        if (empty($resourceList)) {
+            return 0;
+        }
+        $this->indexerProgressHandler->advance(count($resourceList));
+        $result = $this->add($solrCore, $processId, $resourceList);
+
+        if ($result->getStatus() !== 0) {
+            $this->indexerProgressHandler->error(new Exception(
+                $result->getResponse()->getStatusMessage()
+            ));
+            return 0;
+        }
+
+        return count($resourceList);
+    }
+
+    /**
+     * @param string[] $pathList
+     * @return Resource[]|false
+     */
+    private function loadResources(
+        array $pathList,
+        int $offset,
+        int $length
+    ): array|false {
+
+        $maxLength = (count($pathList) - $offset);
+        if ($maxLength <= 0) {
+            return false;
+        }
+
+        if ($length > $maxLength) {
+            $length = $maxLength;
+        }
+
+        $resourceList = [];
+        for ($i = $offset; $i < ($length + $offset); $i++) {
```

Comment:
we can calculate the loop-end beforehand:

```suggestion
        $maxLength = count($pathList) - $offset;
        if ($maxLength <= 0) {
            return false;
        }

        $end = min($length, $maxLength) + $offset;

        $resourceList = [];
        for ($i = $offset; $i < $end; $i++) {
```